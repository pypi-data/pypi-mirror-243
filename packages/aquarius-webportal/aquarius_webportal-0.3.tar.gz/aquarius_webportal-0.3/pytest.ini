[pytest]
log_cli = true
addopts = --doctest-modules --cov-config=.coveragerc --cov-report xml --cov-report term --cov=aquarius_webportal
testpaths = tests