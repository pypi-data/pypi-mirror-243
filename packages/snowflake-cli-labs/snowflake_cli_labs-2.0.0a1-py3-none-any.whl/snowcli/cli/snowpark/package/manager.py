from __future__ import annotations

import logging
import os.path
import tempfile
from pathlib import Path
from shutil import rmtree

from requirements.requirement import Requirement
from snowcli import utils
from snowcli.cli.constants import PACKAGES_DIR
from snowcli.cli.object.stage.manager import StageManager
from snowcli.cli.snowpark.package.utils import (
    CreatedSuccessfully,
    InAnaconda,
    LookupResult,
    NothingFound,
    NotInAnaconda,
    RequiresPackages,
)
from snowcli.utils import SplitRequirements
from snowcli.zipper import zip_dir

log = logging.getLogger(__name__)


def lookup(name: str, install_packages: bool) -> LookupResult:

    package_response = utils.parse_anaconda_packages([Requirement.parse(name)])

    if package_response.snowflake and not package_response.other:
        return InAnaconda(package_response, name)
    elif install_packages:
        status, result = utils.install_packages(
            perform_anaconda_check=True, package_name=name, file_name=None
        )

        if status:
            if result.snowflake:
                return RequiresPackages(result, name)
            else:
                return NotInAnaconda(result, name)

    return NothingFound(SplitRequirements([], []), name)


def upload(file: Path, stage: str, overwrite: bool):
    log.info(f"Uploading {file} to Snowflake @{stage}/{file}...")
    with tempfile.TemporaryDirectory() as temp_dir:
        temp_app_zip_path = utils.prepare_app_zip(file, temp_dir)
        sm = StageManager()
        sm.create(stage)
        put_response = sm.put(temp_app_zip_path, stage, overwrite=overwrite).fetchone()

    message = f"Package {file} {put_response[6]} to Snowflake @{stage}/{file}."

    if put_response[6] == "SKIPPED":
        message = "Package already exists on stage. Consider using --overwrite to overwrite the file."

    return message


def create(zip_name: str):
    file_name = zip_name if zip_name.endswith(".zip") else f"{zip_name}.zip"
    zip_dir(dest_zip=Path(file_name), source=Path.cwd())

    if os.path.exists(file_name):
        return CreatedSuccessfully(zip_name, Path(file_name))


def cleanup_after_install():
    if PACKAGES_DIR.exists():
        rmtree(PACKAGES_DIR)
