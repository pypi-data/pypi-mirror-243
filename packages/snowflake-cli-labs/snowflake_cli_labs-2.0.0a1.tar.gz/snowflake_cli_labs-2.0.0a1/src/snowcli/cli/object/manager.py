from __future__ import annotations

from functools import wraps

from click import ClickException
from snowcli.cli.common.sql_execution import SqlExecutionMixin
from snowcli.cli.constants import OBJECT_TO_NAMES, ObjectNames
from snowflake.connector.cursor import SnowflakeCursor


def _get_object_names(object_name: str) -> ObjectNames:
    object_name = object_name.lower()
    if object_name.lower() not in OBJECT_TO_NAMES:
        raise ClickException(f"Object of type {object_name} is not supported.")
    return OBJECT_TO_NAMES[object_name]


class ObjectManager(SqlExecutionMixin):
    def show(self, *, object_type: str, like: str) -> SnowflakeCursor:
        object_name = _get_object_names(object_type).sf_plural_name
        return self._execute_query(f"show {object_name} like '{like}'")

    def drop(self, *, object_type, name: str) -> SnowflakeCursor:
        object_name = _get_object_names(object_type).sf_name
        return self._execute_query(f"drop {object_name} {name}")

    def describe(self, *, object_type: str, name: str):
        object_name = _get_object_names(object_type).sf_name
        return self._execute_query(f"describe {object_name} {name}")
