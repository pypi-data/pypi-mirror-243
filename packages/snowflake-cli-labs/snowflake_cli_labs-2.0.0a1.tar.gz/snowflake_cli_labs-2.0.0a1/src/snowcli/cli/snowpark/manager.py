from __future__ import annotations

import logging
from typing import Dict, List, Optional

from snowcli.cli.snowpark.common import SnowparkObjectManager
from snowflake.connector.cursor import SnowflakeCursor

log = logging.getLogger(__file__)


class FunctionManager(SnowparkObjectManager):
    @property
    def _object_type(self):
        return "function"

    @property
    def _object_execute(self):
        return "select"

    def create_or_replace(
        self,
        identifier: str,
        return_type: str,
        handler: str,
        artifact_file: str,
        packages: List[str],
        external_access_integrations: Optional[List[str]] = None,
        secrets: Optional[Dict[str, str]] = None,
    ) -> SnowflakeCursor:
        log.debug(f"Creating function {identifier} using @{artifact_file}")
        query = self.create_query(
            identifier,
            return_type,
            handler,
            artifact_file,
            packages,
            external_access_integrations,
            secrets,
        )
        return self._execute_query(query)


class ProcedureManager(SnowparkObjectManager):
    @property
    def _object_type(self):
        return "procedure"

    @property
    def _object_execute(self):
        return "call"

    def create_or_replace(
        self,
        identifier: str,
        return_type: str,
        handler: str,
        artifact_file: str,
        packages: List[str],
        external_access_integrations: Optional[List[str]] = None,
        secrets: Optional[Dict[str, str]] = None,
        execute_as_caller: bool = False,
    ) -> SnowflakeCursor:
        log.debug(f"Creating procedure {identifier} using @{artifact_file}")
        query = self.create_query(
            identifier,
            return_type,
            handler,
            artifact_file,
            packages,
            external_access_integrations,
            secrets,
            execute_as_caller,
        )
        return self._execute_query(query)
