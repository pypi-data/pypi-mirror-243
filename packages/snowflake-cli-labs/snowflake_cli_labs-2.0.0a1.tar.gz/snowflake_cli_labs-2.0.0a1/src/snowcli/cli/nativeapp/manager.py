from __future__ import annotations

import logging
from functools import cached_property
from pathlib import Path
from textwrap import dedent
from typing import Callable, List, Literal, Optional

import jinja2
from click.exceptions import ClickException
from snowcli.cli.common.sql_execution import SqlExecutionMixin
from snowcli.cli.connection.util import make_snowsight_url
from snowcli.cli.nativeapp.artifacts import (
    ArtifactMapping,
    build_bundle,
    translate_artifact,
)
from snowcli.cli.object.stage.diff import (
    DiffResult,
    stage_diff,
    sync_local_diff_with_stage,
)
from snowcli.cli.object.stage.manager import StageManager
from snowcli.cli.project.definition import (
    default_app_package,
    default_application,
    default_role,
)
from snowcli.cli.project.definition_manager import DefinitionManager
from snowcli.cli.project.util import (
    extract_schema,
    to_identifier,
    unquote_identifier,
)
from snowcli.exception import SnowflakeSQLExecutionError
from snowflake.connector import ProgrammingError
from snowflake.connector.cursor import DictCursor

SPECIAL_COMMENT = "GENERATED_BY_SNOWCLI"
LOOSE_FILES_MAGIC_VERSIONS = ["dev_stage", "UNVERSIONED"]

NAME_COL = "name"
COMMENT_COL = "comment"
OWNER_COL = "owner"
VERSION_COL = "version"

ERROR_MESSAGE_2043 = "Object does not exist, or operation cannot be performed."
ERROR_MESSAGE_606 = "No active warehouse selected in the current session."

log = logging.getLogger(__name__)


class ApplicationPackageAlreadyExistsError(ClickException):
    """An application package not created by SnowCLI exists with the same name."""

    def __init__(self, name: str):
        super().__init__(
            f"An Application Package {name} already exists in account that may have been created without snowCLI."
        )


class ApplicationAlreadyExistsError(ClickException):
    """An application not created by SnowCLI exists with the same name."""

    def __init__(self, name: str):
        super().__init__(
            f'A non-dev application "{name}" already exists in the account.'
        )


class CouldNotDropObjectError(ClickException):
    """
    Could not successfully drop the required Snowflake object.
    """

    def __init__(self, message: str):
        super().__init__(message=message)


class UnexpectedOwnerError(ClickException):
    """An operation is blocked becuase an object is owned by an unexpected role."""

    def __init__(self, item: str, expected_owner: str, actual_owner: str):
        super().__init__(
            f"Cannot operate on {item}: owned by {actual_owner} (expected {expected_owner})"
        )


class MissingPackageScriptError(ClickException):
    """A referenced package script was not found."""

    def __init__(self, relpath: str):
        super().__init__(f'Package script "{relpath}" does not exist')


class InvalidPackageScriptError(ClickException):
    """A referenced package script had syntax error(s)."""

    def __init__(self, relpath: str, err: jinja2.TemplateError):
        super().__init__(f'Package script "{relpath}" is not a valid jinja2 template')
        self.err = err


class MissingSchemaError(ClickException):
    """An identifier is missing a schema qualifier."""

    def __init__(self, identifier: str):
        super().__init__(f'Identifier missing a schema qualifier: "{identifier}"')


def find_row(cursor: DictCursor, predicate: Callable[[dict], bool]) -> Optional[dict]:
    """Returns the first row that matches the predicate, or None."""
    return next(
        (row for row in cursor.fetchall() if predicate(row)),
        None,
    )


def _generic_sql_error_handler(
    err: ProgrammingError, role: Optional[str] = None, warehouse: Optional[str] = None
):
    # Potential refactor: If moving away from python 3.8 and 3.9 to >= 3.10, use match ... case
    if err.errno == 2043 or err.msg.__contains__(ERROR_MESSAGE_2043):
        raise ProgrammingError(
            msg=dedent(
                f"""\
                Received error message '{err.msg}' while executing SQL statement.
                '{role}' may not have access to warehouse '{warehouse}'.
                Please grant usage privilege on warehouse to this role.
                """
            ),
            errno=err.errno,
        )
    elif err.errno == 606 or err.msg.__contains__(ERROR_MESSAGE_606):
        raise ProgrammingError(
            msg=dedent(
                f"""\
                Received error message '{err.msg}' while executing SQL statement.
                Please provide a warehouse for the active session role in your project definition file, config.toml file, or via command line.
                """
            ),
            errno=err.errno,
        )
    raise err


class NativeAppManager(SqlExecutionMixin):
    definition_manager: DefinitionManager

    def __init__(self, search_path: Optional[str] = None):
        super().__init__()
        self.definition_manager = DefinitionManager(search_path or "")

    @property
    def project_root(self) -> Path:
        return self.definition_manager.project_root

    @property
    def definition(self) -> dict:
        return self.definition_manager.project_definition["native_app"]

    @cached_property
    def artifacts(self) -> List[ArtifactMapping]:
        return [translate_artifact(item) for item in self.definition["artifacts"]]

    @cached_property
    def deploy_root(self) -> Path:
        return Path(self.project_root, self.definition["deploy_root"])

    @cached_property
    def package_scripts(self) -> List[str]:
        """
        Relative paths to package scripts from the project root.
        """
        return self.definition.get("package", {}).get("scripts", [])

    @cached_property
    def stage_fqn(self) -> str:
        return f'{self.package_name}.{self.definition["source_stage"]}'

    @cached_property
    def stage_schema(self) -> Optional[str]:
        return extract_schema(self.stage_fqn)

    @cached_property
    def package_warehouse(self) -> Optional[str]:
        return self.definition.get("package", {}).get("warehouse", self._conn.warehouse)

    @cached_property
    def application_warehouse(self) -> Optional[str]:
        return self.definition.get("application", {}).get(
            "warehouse", self._conn.warehouse
        )

    @cached_property
    def project_identifier(self) -> str:
        # name is expected to be a valid Snowflake identifier, but PyYAML
        # will sometimes strip out double quotes so we try to get them back here.
        return to_identifier(self.definition["name"])

    @cached_property
    def package_name(self) -> str:
        return to_identifier(
            self.definition.get("package", {}).get(
                "name", default_app_package(self.project_identifier)
            )
        )

    @cached_property
    def package_role(self) -> str:
        return self.definition.get("package", {}).get("role", None) or default_role()

    @cached_property
    def app_name(self) -> str:
        return to_identifier(
            self.definition.get("application", {}).get(
                "name", default_application(self.project_identifier)
            )
        )

    @cached_property
    def app_role(self) -> str:
        return (
            self.definition.get("application", {}).get("role", None) or default_role()
        )

    @cached_property
    def debug_mode(self) -> bool:
        return self.definition.get("application", {}).get("debug", True)

    def build_bundle(self) -> None:
        """
        Populates the local deploy root from artifact sources.
        """
        build_bundle(self.project_root, self.deploy_root, self.artifacts)

    def sync_deploy_root_with_stage(self, role: str) -> DiffResult:
        """
        Ensures that the files on our remote stage match the artifacts we have in
        the local filesystem. Returns the DiffResult used to make changes.
        """

        # Does a stage already exist within the app pkg, or we need to create one?
        # Using "if not exists" should take care of either case.
        log.info("Checking if stage exists, or creating a new one if none exists.")
        with self.use_role(role):
            self._execute_query(
                f"create schema if not exists {self.package_name}.{self.stage_schema}"
            )
            self._execute_query(
                f"""
                    create stage if not exists {self.stage_fqn}
                    encryption = (TYPE = 'SNOWFLAKE_SSE')"""
            )

        # Perform a diff operation and display results to the user for informational purposes
        log.info(
            f'Performing a diff between the Snowflake stage and your local deploy_root ("{self.deploy_root}") directory.'
        )
        diff: DiffResult = stage_diff(self.deploy_root, self.stage_fqn)
        log.info("Listing results of diff:")
        log.info(f"New files only on your local: {','.join(diff.only_local)}")
        log.info(f"New files only on the stage: {','.join(diff.only_on_stage)}")
        log.info(
            f"Existing files modified or status unknown: {','.join(diff.different)}"
        )
        log.info(f"Existing files identical to the stage: {','.join(diff.identical)}")

        # Upload diff-ed files to app pkg stage
        if diff.has_changes():
            log.info(
                f"Uploading diff-ed files from your local {self.deploy_root} directory to the Snowflake stage."
            )
            sync_local_diff_with_stage(
                role=role,
                deploy_root_path=self.deploy_root,
                diff_result=diff,
                stage_path=self.stage_fqn,
            )
        return diff

    def _apply_package_scripts(self) -> None:
        """
        Assuming the application package exists and we are using the correct role,
        applies all package scripts in-order to the application package.
        """
        env = jinja2.Environment(
            loader=jinja2.loaders.FileSystemLoader(self.project_root),
            keep_trailing_newline=True,
            undefined=jinja2.StrictUndefined,
        )

        queued_queries = []
        for relpath in self.package_scripts:
            try:
                template = env.get_template(relpath)
                result = template.render(dict(package_name=self.package_name))
                queued_queries.append(result)

            except jinja2.TemplateNotFound as e:
                raise MissingPackageScriptError(e.name)

            except jinja2.TemplateSyntaxError as e:
                raise InvalidPackageScriptError(e.name, e)

            except jinja2.UndefinedError as e:
                raise InvalidPackageScriptError(relpath, e)

        # once we're sure all the templates expanded correctly, execute all of them
        try:
            if self.package_warehouse:
                self._execute_query(f"use warehouse {self.package_warehouse}")

            for i, queries in enumerate(queued_queries):
                log.info(f"Applying package script: {self.package_scripts[i]}")
                self._execute_queries(queries)
        except ProgrammingError as err:
            _generic_sql_error_handler(
                err, role=self.package_role, warehouse=self.package_warehouse
            )

    def _create_dev_app(self, diff: DiffResult) -> None:
        """
        (Re-)creates the application with our up-to-date stage.
        """
        with self.use_role(self.app_role):
            try:
                if self.application_warehouse:
                    self._execute_query(f"use warehouse {self.application_warehouse}")
            except ProgrammingError as err:
                _generic_sql_error_handler(
                    err=err, role=self.app_role, warehouse=self.application_warehouse
                )

            show_app_cursor = self._execute_query(
                f"show applications like '{unquote_identifier(self.app_name)}'",
                cursor_class=DictCursor,
            )

            # There can only be one possible pre-existing app with the same name
            show_app_row = find_row(
                show_app_cursor,
                lambda row: row[NAME_COL] == unquote_identifier(self.app_name),
            )

            if show_app_row is not None:
                if show_app_row[COMMENT_COL] != SPECIAL_COMMENT or (
                    show_app_row[VERSION_COL] not in LOOSE_FILES_MAGIC_VERSIONS
                ):
                    raise ApplicationAlreadyExistsError(self.app_name)

                actual_owner = show_app_row[OWNER_COL]
                if actual_owner != unquote_identifier(self.app_role):
                    raise UnexpectedOwnerError(
                        self.app_name, self.app_role, actual_owner
                    )

                try:
                    if diff.has_changes():
                        # the app needs to be upgraded
                        log.info(f"Upgrading existing application {self.app_name}.")
                        self._execute_query(
                            f"alter application {self.app_name} upgrade using @{self.stage_fqn}"
                        )

                    # ensure debug_mode is up-to-date
                    self._execute_query(
                        f"alter application {self.app_name} set debug_mode = {self.debug_mode}"
                    )
                    return
                except ProgrammingError as err:
                    _generic_sql_error_handler(err)

            # Create an app using "loose files" / stage dev mode.
            log.info(f"Creating new application {self.app_name} in account.")

            if self.app_role != self.package_role:
                with self.use_role(new_role=self.package_role):
                    self._execute_queries(
                        dedent(
                            f"""\
                        grant install, develop on application package {self.package_name} to role {self.app_role};
                        grant usage on schema {self.package_name}.{self.stage_schema} to role {self.app_role};
                        grant read on stage {self.stage_fqn} to role {self.app_role};
                        """
                        )
                    )

            stage_name = StageManager.quote_stage_name(self.stage_fqn)

            try:
                self._execute_query(
                    f"""
                    create application {self.app_name}
                        from application package {self.package_name}
                        using {stage_name}
                        debug_mode = {self.debug_mode}
                        comment = {SPECIAL_COMMENT}
                    """,
                )
            except ProgrammingError as err:
                _generic_sql_error_handler(err)

    def app_exists(self) -> bool:
        """Returns True iff the application exists on Snowflake."""
        with self.use_role(self.app_role):
            show_app_cursor = self._execute_query(
                f"show applications like '{unquote_identifier(self.app_name)}'",
                cursor_class=DictCursor,
            )
            return (
                find_row(
                    show_app_cursor,
                    lambda row: row[NAME_COL] == unquote_identifier(self.app_name),
                )
                is not None
            )

    def app_run(self) -> None:
        """
        Implementation of the "snow app run" dev command.
        """
        with self.use_role(self.package_role):
            show_cursor = self._execute_query(
                f"show application packages like '{unquote_identifier(self.package_name)}'",
                cursor_class=DictCursor,
            )

            if show_cursor.rowcount is None:
                raise SnowflakeSQLExecutionError()

            # There can be maximum one possible pre-existing app pkg with the same name
            show_app_row = find_row(
                show_cursor,
                lambda row: row[NAME_COL] == unquote_identifier(self.package_name),
            )
            if show_app_row is None:
                # Create an app pkg, with distribution = internal to avoid triggering security scan
                log.info(
                    f"Creating new application package {self.package_name} in account."
                )
                self._execute_query(
                    f"""
                    create application package {self.package_name}
                        comment = {SPECIAL_COMMENT}
                        distribution = internal
                    """
                )
            else:
                row_comment = show_app_row[
                    COMMENT_COL
                ]  # Because we use a DictCursor, we are guaranteed a dictionary instead of indexing through a list.

                if row_comment != SPECIAL_COMMENT:
                    raise ApplicationPackageAlreadyExistsError(self.package_name)

                actual_owner = show_app_row[OWNER_COL]
                if actual_owner != unquote_identifier(self.package_role):
                    raise UnexpectedOwnerError(
                        self.app_name, self.app_role, actual_owner
                    )

            # now that the application package exists, create shared data
            self._apply_package_scripts()

            # Upload files from deploy root local folder to the above stage
            diff = self.sync_deploy_root_with_stage(self.package_role)

        self._create_dev_app(diff)

    def get_snowsight_url(self) -> str:
        """Returns the URL that can be used to visit this app via Snowsight."""
        name = unquote_identifier(self.app_name)
        return make_snowsight_url(self._conn, f"/#/apps/application/{name}")

    def drop_object(
        self,
        object_name: str,
        object_role: str,
        object_type: Literal["application", "package"],
        query_dict: dict,
    ) -> None:
        """
        N.B. query_dict['show'] must be a like % clause
        """
        log_object_type = (
            "Application Package" if object_type == "package" else object_type
        )

        with self.use_role(object_role):
            show_obj_query = f"{query_dict['show']} '{unquote_identifier(object_name)}'"
            show_obj_cursor = self._execute_query(
                show_obj_query, cursor_class=DictCursor
            )

            if show_obj_cursor.rowcount is None:
                raise SnowflakeSQLExecutionError(show_obj_query)

            show_obj_row = find_row(
                show_obj_cursor,
                lambda row: row[NAME_COL] == unquote_identifier(object_name),
            )
            if show_obj_row is None:
                raise CouldNotDropObjectError(
                    f"Role {object_role} does not own any {log_object_type.lower()} with the name {object_name}!"
                )

            # There can only be one possible pre-existing object with the same name
            row_comment = show_obj_row[
                COMMENT_COL
            ]  # Because we use a DictCursor, we are guaranteed a dictionary instead of indexing through a list.

            if row_comment != SPECIAL_COMMENT:
                raise CouldNotDropObjectError(
                    f"{log_object_type} {object_name} was not created by SnowCLI. Cannot drop the {log_object_type.lower()}."
                )

            log.info(f"Dropping {log_object_type.lower()} {object_name} now.")
            drop_query = f"{query_dict['drop']} {object_name}"
            try:
                self._execute_query(drop_query)
            except:
                # Case if an object exists but owned by a different role.
                raise SnowflakeSQLExecutionError(drop_query)
            log.info(f"Dropped {log_object_type.lower()} {object_name} successfully.")

    def teardown(self) -> None:
        # Drop the application first
        self.drop_object(
            object_name=self.app_name,
            object_role=self.app_role,
            object_type="application",
            query_dict={"show": "show applications like", "drop": "drop application"},
        )

        # Drop the application package next
        self.drop_object(
            object_name=self.package_name,
            object_role=self.package_role,
            object_type="package",
            query_dict={
                "show": "show application packages like",
                "drop": "drop application package",
            },
        )
