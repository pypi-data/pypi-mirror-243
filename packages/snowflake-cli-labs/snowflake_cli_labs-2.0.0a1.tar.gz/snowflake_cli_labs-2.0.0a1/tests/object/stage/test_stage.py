from pathlib import Path
from tempfile import TemporaryDirectory
from unittest import mock

import pytest
from snowcli.cli.object.stage.manager import StageManager
from snowflake.connector.cursor import DictCursor

STAGE_MANAGER = "snowcli.cli.object.stage.manager.StageManager"


@mock.patch(f"{STAGE_MANAGER}._execute_query")
def test_stage_list(mock_execute, runner, mock_cursor):
    mock_execute.return_value = mock_cursor(["row"], [])
    result = runner.invoke(["object", "stage", "list", "-c", "empty", "stageName"])
    assert result.exit_code == 0, result.output
    mock_execute.assert_called_once_with("ls @stageName")


@mock.patch(f"{STAGE_MANAGER}._execute_query")
def test_stage_list_quoted(mock_execute, runner, mock_cursor):
    mock_execute.return_value = mock_cursor(["row"], [])
    result = runner.invoke(["object", "stage", "list", "-c", "empty", '"stage name"'])
    assert result.exit_code == 0, result.output
    mock_execute.assert_called_once_with("ls '@\"stage name\"'")


@mock.patch(f"{STAGE_MANAGER}._execute_query")
def test_stage_copy_remote_to_local(mock_execute, runner, mock_cursor):
    mock_execute.return_value = mock_cursor(["row"], [])
    with TemporaryDirectory() as tmp_dir:
        result = runner.invoke(
            ["object", "stage", "copy", "-c", "empty", "@stageName", str(tmp_dir)]
        )
    assert result.exit_code == 0, result.output
    mock_execute.assert_called_once_with(
        f"get @stageName file://{Path(tmp_dir).resolve()}/ parallel=4"
    )


@mock.patch(f"{STAGE_MANAGER}._execute_query")
def test_stage_copy_remote_to_local_quoted(mock_execute, runner, mock_cursor):
    mock_execute.return_value = mock_cursor(["row"], [])
    with TemporaryDirectory() as tmp_dir:
        result = runner.invoke(
            ["object", "stage", "copy", "-c", "empty", '@"stage name"', str(tmp_dir)]
        )
    assert result.exit_code == 0, result.output
    mock_execute.assert_called_once_with(
        f"get '@\"stage name\"' file://{Path(tmp_dir).resolve()}/ parallel=4"
    )


@mock.patch(f"{STAGE_MANAGER}._execute_query")
def test_stage_copy_local_to_remote(mock_execute, runner, mock_cursor):
    mock_execute.return_value = mock_cursor(["row"], [])
    with TemporaryDirectory() as tmp_dir:
        result = runner.invoke(
            [
                "object",
                "stage",
                "copy",
                "-c",
                "empty",
                "--overwrite",
                "--parallel",
                42,
                str(tmp_dir),
                "@stageName",
            ]
        )
    assert result.exit_code == 0, result.output
    mock_execute.assert_called_once_with(
        f"put file://{Path(tmp_dir).resolve()}/* @stageName auto_compress=false parallel=42 overwrite=True"
    )


@mock.patch(f"{STAGE_MANAGER}._execute_query")
def test_stage_copy_local_to_remote_quoted(mock_execute, runner, mock_cursor):
    mock_execute.return_value = mock_cursor(["row"], [])
    with TemporaryDirectory() as tmp_dir:
        result = runner.invoke(
            [
                "object",
                "stage",
                "copy",
                "-c",
                "empty",
                "--overwrite",
                "--parallel",
                42,
                str(tmp_dir),
                '@"stage name"',
            ]
        )
    assert result.exit_code == 0, result.output
    mock_execute.assert_called_once_with(
        f"put file://{Path(tmp_dir).resolve()}/* '@\"stage name\"' auto_compress=false parallel=42 overwrite=True"
    )


@mock.patch(f"{STAGE_MANAGER}._execute_query")
def test_stage_copy_local_to_remote_star(mock_execute, runner, mock_cursor):
    mock_execute.return_value = mock_cursor(["row"], [])
    with TemporaryDirectory() as tmp_dir:
        result = runner.invoke(
            [
                "object",
                "stage",
                "copy",
                "-c",
                "empty",
                "--overwrite",
                "--parallel",
                42,
                str(tmp_dir) + "/*.py",
                "@stageName",
            ]
        )
    assert result.exit_code == 0, result.output
    mock_execute.assert_called_once_with(
        f"put file://{Path(tmp_dir).resolve()}/*.py @stageName auto_compress=false parallel=42 overwrite=True"
    )


@pytest.mark.parametrize(
    "source, dest",
    [
        ("@snow/stage", "@stage/snow"),
        ("snow://stage", "snow://stage/snow"),
        ("local/path", "other/local/path"),
    ],
)
def test_copy_throws_error_for_same_platform_operation(runner, source, dest, snapshot):
    result = runner.invoke(["object", "stage", "copy", source, dest])
    assert result.exit_code == 1
    assert result.output == snapshot


@mock.patch(f"{STAGE_MANAGER}._execute_query")
def test_stage_create(mock_execute, runner, mock_cursor):
    mock_execute.return_value = mock_cursor(["row"], [])
    result = runner.invoke(["object", "stage", "create", "-c", "empty", "stageName"])
    assert result.exit_code == 0, result.output
    mock_execute.assert_called_once_with("create stage if not exists stageName")


@mock.patch(f"{STAGE_MANAGER}._execute_query")
def test_stage_create_quoted(mock_execute, runner, mock_cursor):
    mock_execute.return_value = mock_cursor(["row"], [])
    result = runner.invoke(["object", "stage", "create", "-c", "empty", '"stage name"'])
    assert result.exit_code == 0, result.output
    mock_execute.assert_called_once_with('create stage if not exists "stage name"')


@mock.patch("snowcli.cli.object.commands.ObjectManager._execute_query")
def test_stage_drop(mock_execute, runner, mock_cursor):
    mock_execute.return_value = mock_cursor(["row"], [])
    result = runner.invoke(["object", "drop", "stage", "stageName", "-c", "empty"])
    assert result.exit_code == 0, result.output
    mock_execute.assert_called_once_with("drop stage stageName")


@mock.patch("snowcli.cli.object.commands.ObjectManager._execute_query")
def test_stage_drop_quoted(mock_execute, runner, mock_cursor):
    mock_execute.return_value = mock_cursor(["row"], [])
    result = runner.invoke(["object", "drop", "stage", '"stage name"', "-c", "empty"])
    assert result.exit_code == 0, result.output
    mock_execute.assert_called_once_with('drop stage "stage name"')


@mock.patch(f"{STAGE_MANAGER}._execute_query")
def test_stage_remove(mock_execute, runner, mock_cursor):
    mock_execute.return_value = mock_cursor(["row"], [])
    result = runner.invoke(
        ["object", "stage", "remove", "-c", "empty", "stageName", "my/file/foo.csv"]
    )
    assert result.exit_code == 0, result.output
    mock_execute.assert_called_once_with("remove @stageName/my/file/foo.csv")


@mock.patch(f"{STAGE_MANAGER}._execute_query")
def test_stage_remove_quoted(mock_execute, runner, mock_cursor):
    mock_execute.return_value = mock_cursor(["row"], [])
    result = runner.invoke(
        ["object", "stage", "remove", "-c", "empty", '"stage name"', "my/file/foo.csv"]
    )
    assert result.exit_code == 0, result.output
    mock_execute.assert_called_once_with("remove '@\"stage name\"/my/file/foo.csv'")


@mock.patch(f"{STAGE_MANAGER}._execute_query")
def test_stage_internal_remove(mock_execute, mock_cursor):
    mock_execute.return_value = mock_cursor([{"CURRENT_ROLE()": "old_role"}], [])
    sm = StageManager()
    sm._remove("stageName", "my/file/foo.csv", "new_role")
    expected = [
        mock.call("select current_role()", cursor_class=DictCursor),
        mock.call("use role new_role"),
        mock.call("remove @stageName/my/file/foo.csv"),
        mock.call("use role old_role"),
    ]
    assert mock_execute.mock_calls == expected


@mock.patch(f"{STAGE_MANAGER}._execute_query")
def test_stage_internal_remove_quoted(mock_execute, mock_cursor):
    mock_execute.return_value = mock_cursor([{"CURRENT_ROLE()": "old_role"}], [])
    sm = StageManager()
    sm._remove('"stage name"', "my/file/foo.csv", "new_role")
    expected = [
        mock.call("select current_role()", cursor_class=DictCursor),
        mock.call("use role new_role"),
        mock.call("remove '@\"stage name\"/my/file/foo.csv'"),
        mock.call("use role old_role"),
    ]
    assert mock_execute.mock_calls == expected


@mock.patch(f"{STAGE_MANAGER}._execute_query")
def test_stage_internal_remove_no_role_change(mock_execute, mock_cursor):
    mock_execute.return_value = mock_cursor([{"CURRENT_ROLE()": "old_role"}], [])
    sm = StageManager()
    sm._remove("stageName", "my/file/foo.csv", "old_role")
    expected = [
        mock.call("select current_role()", cursor_class=DictCursor),
        mock.call("remove @stageName/my/file/foo.csv"),
    ]
    assert mock_execute.mock_calls == expected


@mock.patch(f"{STAGE_MANAGER}._execute_query")
def test_stage_internal_put(mock_execute, mock_cursor):
    mock_execute.return_value = mock_cursor([{"CURRENT_ROLE()": "old_role"}], [])
    with TemporaryDirectory() as tmp_dir:
        sm = StageManager()
        sm._put(Path(tmp_dir).resolve(), "stageName", "new_role")
        expected = [
            mock.call("select current_role()", cursor_class=DictCursor),
            mock.call("use role new_role"),
            mock.call(
                f"put file://{Path(tmp_dir).resolve()} @stageName auto_compress=false parallel=4 overwrite=False"
            ),
            mock.call("use role old_role"),
        ]
        assert mock_execute.mock_calls == expected


@mock.patch(f"{STAGE_MANAGER}._execute_query")
def test_stage_internal_put_quoted(mock_execute, mock_cursor):
    mock_execute.return_value = mock_cursor([{"CURRENT_ROLE()": "old_role"}], [])
    with TemporaryDirectory() as tmp_dir:
        sm = StageManager()
        sm._put(Path(tmp_dir).resolve(), '"stage name"', "new_role")
        expected = [
            mock.call("select current_role()", cursor_class=DictCursor),
            mock.call("use role new_role"),
            mock.call(
                f"put file://{Path(tmp_dir).resolve()} '@\"stage name\"' auto_compress=false parallel=4 overwrite=False"
            ),
            mock.call("use role old_role"),
        ]
        assert mock_execute.mock_calls == expected
