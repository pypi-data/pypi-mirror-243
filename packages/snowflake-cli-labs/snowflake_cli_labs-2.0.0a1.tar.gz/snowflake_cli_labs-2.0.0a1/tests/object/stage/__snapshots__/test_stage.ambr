# serializer version: 1
# name: test_copy_throws_error_for_same_platform_operation[@snow/stage-@stage/snow]
  '''
  ╭─ Error ──────────────────────────────────────────────────────────────────────╮
  │ Both source and target path are remote. This operation is not supported.     │
  ╰──────────────────────────────────────────────────────────────────────────────╯
  
  '''
# ---
# name: test_copy_throws_error_for_same_platform_operation[local/path-other/local/path]
  '''
  ╭─ Error ──────────────────────────────────────────────────────────────────────╮
  │ Both source and target path are local. This operation is not supported.      │
  ╰──────────────────────────────────────────────────────────────────────────────╯
  
  '''
# ---
# name: test_copy_throws_error_for_same_platform_operation[snow://stage-snow://stage/snow]
  '''
  ╭─ Error ──────────────────────────────────────────────────────────────────────╮
  │ Both source and target path are remote. This operation is not supported.     │
  ╰──────────────────────────────────────────────────────────────────────────────╯
  
  '''
# ---
